commands/filescommands/objectcommands/uploaddir.go

```diff
@@ -0,0 +1,235 @@
+
+	opts := objects.CreateOpts{
+		ContentLength: int64(c.Int("content-length")),
+		ContentType:   c.String("content-type"),
+	}
+
+	if c.IsSet("content-encoding") && c.String("content-encoding") != "gzip" {
+		opts.ContentEncoding = c.String("content-encoding")
+	}
+
+	conc := c.Int("concurrency")
+	if conc <= 0 {
+		conc = 5
+	}
+
+	resource.Params = &paramsUploadDir{
+		container:   c.String("container"),
+		dir:         c.String("dir"),
+		opts:        opts,
+		concurrency: c.Int("concurrency"),
+		quiet:       c.Bool("quiet"),
+	}
+
+	return nil
+}
+
+func (command *commandUploadDir) StdinField() string {
+	return "dir"
+}
+
+func (command *commandUploadDir) HandlePipe(resource *handler.Resource, item string) error {
+	resource.Params.(*paramsUploadDir).dir = item
+	return nil
+}
+
+func (command *commandUploadDir) HandleSingle(resource *handler.Resource) error {
+	err := command.Ctx.CheckFlagsSet([]string{"dir"})
+	if err != nil {
+		return err
+	}
+	resource.Params.(*paramsUploadDir).dir = command.Ctx.CLIContext.String("dir")
+	return nil
+}
+
+func (command *commandUploadDir) Execute(resource *handler.Resource) {
+	// bump thread count to number of available CPUs
+	runtime.GOMAXPROCS(runtime.NumCPU())
+
+	params := resource.Params.(*paramsUploadDir)
+	jobs := make(chan string)
+	results := make(chan *handler.Resource)
+
+	var wg sync.WaitGroup
+	var totalSize uint64
+	start := time.Now()
+
+	go func() {
+		for r := range results {
+			command.Ctx.WaitGroup.Add(1)
+			command.Ctx.Results <- r
+		}
+		command.Ctx.WaitGroup.Wait()
+	}()
+
+	for i := 0; i < params.concurrency; i++ {
+		wg.Add(1)
+		go func(totalSize *uint64) {
+			for p := range jobs {
+				var re *handler.Resource
+
+				ticker := backoff.NewTicker(backoff.NewExponentialBackOff())
+				for _ = range ticker.C {
+					re = command.handle(p, params)
+					if re.Err != nil {
+						continue
+					}
+
+					ticker.Stop()
+					break
+				}
+
+				fi, err := os.Stat(p)
+				if err == nil {
+					*totalSize += uint64(fi.Size())
+				}
+
+				results <- re
+			}
+			wg.Done()
+		}(&totalSize)
+	}
+
+	filepath.Walk(params.dir, func(path string, info os.FileInfo, err error) error {
+		if !info.IsDir() {
+			jobs <- path
+		}
+		return nil
+	})
+	close(jobs)
+
+	wg.Wait()
+	close(results)
+
+	resource.Result = fmt.Sprintf("Finished! %s uploaded in %s\n", humanize.Bytes(totalSize), humanize.RelTime(start, time.Now(), "", ""))
+}
+
+func (command *commandUploadDir) handle(p string, params *paramsUploadDir) *handler.Resource {
+	re := &handler.Resource{}
+
+	file, err := os.Open(p)
+	defer file.Close()
+
+	if err != nil {
+		re.Err = err
+		return re
+	}
+
+	on := strings.TrimPrefix(p, params.dir+"/")
+	res := objects.Create(command.Ctx.ServiceClient, params.container, on, file, params.opts)
+	re.Err = res.Err
+
+	if res.Err == nil {
+		if params.quiet == true {
+			re.Result = ""
+		} else {
+			re.Result = fmt.Sprintf("Uploaded %s to %s\n", on, params.container)
```

Comment:
yep, this was also mentioned, but @kenperkins thought it would be a better idea to implement this at a lower level (maybe in gophercloud itself). I think it's probably better since there might be other scenarios where we need to ensure transmission integrity. I actually implemented a checksum comparison for this PR but stashed it. I can pop everything back onto this PR if we think it's the right place